### PR TITLE
PTCD-676 Added GetAddressAddressSearchService

### PR DESCRIPTION
**Pull request by jameslawesynergy**
Genericised `LoqateAddressSearch` namespace
Added `GetAddressAddressSearchService` implementation of `IAddressSearchService`

### CODE REVIEW

**Review by timabell — APPROVED** [review]

File: src/Dfc.CourseDirectory.WebV2/AddressSearch/GetAddressAddressSearchService.cs [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        public async Task<AddressDetail> GetById(string id)
+        {
+            if (string.IsNullOrWhiteSpace(id))
+            {
+                throw new ArgumentException($"{nameof(id)} cannot be null or whitespace.");
+            }
+
+            var idSegments = id.Split(IdDelimiter);
+
+            var results = await FindAddresses(idSegments?.FirstOrDefault() ?? throw new ArgumentException($"Failed to parse postcode from {nameof(id)}."));
```
Comment (gunndabad) [inline src/Dfc.CourseDirectory.WebV2/AddressSearch/GetAddressAddressSearchService.cs:34]:
```suggestion
            var results = await FindAddresses(idSegments.FirstOrDefault() ?? throw new ArgumentException($"Failed to parse postcode from {nameof(id)}."));
```

Comment (gunndabad) [reply src/Dfc.CourseDirectory.WebV2/AddressSearch/GetAddressAddressSearchService.cs:34]: Will `idSegments?.FirstOrDefault()` ever be null?

File: src/Dfc.CourseDirectory.WebV2/AddressSearch/GetAddressAddressSearchService.cs [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+            if (results == null)
+            {
+                return null;
+            }
+
+            var result = results.Addresses.FirstOrDefault(i => i.Id == (idSegments?.Skip(1).FirstOrDefault() ?? throw new ArgumentException($"Failed to parse id from {nameof(id)}")));
```
Comment (gunndabad) [inline src/Dfc.CourseDirectory.WebV2/AddressSearch/GetAddressAddressSearchService.cs:41]:
```suggestion
            var result = results.Addresses.FirstOrDefault(i => i.Id == (idSegments.Skip(1).FirstOrDefault() ?? throw new ArgumentException($"Failed to parse id from {nameof(id)}")));
```

**Review by soharadesign — APPROVED** [review]

File: src/Dfc.CourseDirectory.WebV2/AddressSearch/GetAddressAddressSearchService.cs [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+namespace Dfc.CourseDirectory.WebV2.AddressSearch
+{
+    public class GetAddressAddressSearchService : IAddressSearchService
+    {
+        private const string IdDelimiter = "::";
+
+        private readonly HttpClient _httpClient;
+        private readonly Options _options;
+
+        public GetAddressAddressSearchService(HttpClient httpClient, Options options)
```
Comment (gunndabad) [inline src/Dfc.CourseDirectory.WebV2/AddressSearch/GetAddressAddressSearchService.cs:19]: If `Options` is specific to `GetAddress` then I think its name should indicate that

File: src/Dfc.CourseDirectory.WebV2/AddressSearch/GetAddressAddressSearchService.cs (hunk 1 above)
Comment (jameslawesynergy) [reply src/Dfc.CourseDirectory.WebV2/AddressSearch/GetAddressAddressSearchService.cs:34]: No, you're quite right.

File: src/Dfc.CourseDirectory.WebV2/AddressSearch/GetAddressAddressSearchService.cs (hunk 3 above)
Comment (jameslawesynergy) [reply src/Dfc.CourseDirectory.WebV2/AddressSearch/GetAddressAddressSearchService.cs:19]: Agreed

File: src/Dfc.CourseDirectory.WebV2/AddressSearch/GetAddressAddressSearchService.cs [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+            {
+                return null;
+            }
+
+            if (string.IsNullOrWhiteSpace(id))
+            {
+                return null;
+            }
+
+            var results = await FindAddresses(idSegments.FirstOrDefault());
```
[the remark this answers is not in the document]
Comment (jameslawesynergy) [reply src/Dfc.CourseDirectory.WebV2/AddressSearch/GetAddressAddressSearchService.cs:46]: Thank you. Rushing through this and not paying attention!

**Review by gunndabad — APPROVED** [review]

**Pull request merged**